### PR TITLE
Unit tests pass with both Python 2.7.11 and 3.5.0

### PR DESCRIPTION
**Pull request by pineapplemachine**
Fix issues introduced by https://github.com/humangeo/DateSense/pull/1

- Package is again functional in Python 2.7
- Unit tests no longer produce a syntax error in Python 3

### CODE REVIEW

File: DateSense/DSoptions.py [hunk 1]
```diff
@@ -300,7 +300,7 @@ def initialize(self, dates):
             the formatting should be detected.
         '''
         # If it's just one string, turn it into a collection like the methods expect
-        if isinstance(dates, str):
+        if isinstance(dates, ("".__class__, u"".__class__)):
```
Comment (dwyerk) [inline DateSense/DSoptions.py:303]: This is a clever hack that someone on Stackoverflow came up with, but I think explicit is better than implicit. Code golf is not necessary here.

**Review by dwyerk — DISMISSED** [review]

File: DateSense/DSoptions.py (hunk 1 above)
Comment (pineapplemachine) [reply DateSense/DSoptions.py:303]: There is no more explicit way to perform this check without adding a new external dependency such as `six` or `future`. The alternative to this and to adding another dependency is to catch a `NameError` when `basestring` is not defined in Python 3 and check for `str` instead. The `NameError` solution would be unclear to anyone who was not familiar with why checking if things are strings differs in Python 2 and 3. This solution is very clear - Is this object an instance of the same class that string literals are instances of?

Comment (sfairgrieve) [reply DateSense/DSoptions.py:303]: @pineapplemachine correct me if I'm wrong, but I think the main reason for @dwyerk's change request is ensuring full Python 2 support.  I get that it works in Python 3 and most of the time in Python 2, but isn't `isinstance(dates, str)` insufficient for checking if something is a string in Python 2?  With the current code, someone using Python 2 could pass in `u'2018-01-01T00:00:00Z'` as the `dates` variable and the `isinstance(dates, str)` check would return `False` when we actually want it to return `True`.  

An alternative approach might be to check if `basestring` is defined, and if not, set it equal to `str`, then keep this line unchanged.  I don't think that's better than @dwyerk's suggestion, though. 
 Thoughts?

Comment (pineapplemachine) [reply DateSense/DSoptions.py:303]: > With the current code, someone using Python 2 could pass in `u'2018-01-01T00:00:00Z'` as the `dates` variable and the `isinstance(dates, str)` check would return `False` when we actually want it to return `True`.

@sfairgrieve, that is not correct. Here is an example from my Python 2.7.9 shell.

``` python
>>> isinstance("binary string", ("".__class__, u"".__class__))
True
>>> isinstance(u"unicode string", ("".__class__, u"".__class__))
True
```

Here is an example of DateSense accepting a unicode string when running with Python 2.7 (using the PR branch):

``` python
format = DateSense.detect_format(u"2018-04-26").get_format_string()
print(format) # correctly prints u"%Y-%m-%d"
```

Comment (sfairgrieve) [reply DateSense/DSoptions.py:303]: @pineapplemachine You're right.  I misread the comment/history here.  Disregard.

Comment (pineapplemachine) [reply DateSense/DSoptions.py:303]: I also after re-reading see that you may have meant something different from saying that the PR change would fail for unicode inputs in Python 2.7? Confusion all around, I guess.

**Comment by dwyerk** [comment]
Your solution is not my preferred solution, and I don't appreciate the combative tone you've taken with your DMs to me. Since you have no interest in incorporating feedback and there are no production users of this code. I'll just accept your changes so that you will stop sending me angry messages.

**Comment by pineapplemachine** [comment]
I apologize for commenting here, contrary to my claim in the most recent angry message that I was disengaging, but I wanted to say thanks for merging the fix and to hopefully clarify the situation for anyone passing by who is not privy to the DMs in which you accused me of bullying and aggression and issued insults. (When I said I was disengaging, I had not expected you to follow up your private insults with publicly-visible ones!)

Thanks for finally merging. :+1:

**Comment by dwyerk** [comment]
Thanks for pointing out that we need to have a code of conduct for any
contributors.

**Comment by pineapplemachine** [comment]
@sfairgrieve - In the angry messages which Kevin mentioned, I said that I should be given write access to this repository. If not, I will maintain a fork instead. However, if this must be the case, then the license will not permit the maintained fork to be used in HumanGeo's commercial or closed-source projects; this means that HumanGeo will no longer be able to benefit from my knowledge of or labor on DateSense to support its products without negotiating a paid software license. Additionally, if I am not given write access, then the next time that someone unwittingly breaks the repository like what just happened a week ago, you will not learn about it from me. (There is not currently anyone else qualified to confidently determine whether a given change to this repository is safe or not.)

It will be easier for me and more beneficial to HumanGeo if I am able to maintain this original repository.

This was my project and I feel responsible for it. Except for the small prior PR which this one is intended to fix, I have been the sole contributor to this project. I am insisting on write access because it is not acceptable that I should face obstruction and harassment from a HumanGeo employee when I attempt to maintain it.

In this situation, Kevin abused his position of power to a confounding extent. He seems to believe that he has been victimized by my insistence that I be allowed to fix his mistake. When he finally relented, he claimed that the issue was that I would not accept feedback - this is in reference to the same comment which you saw was not constructive or valid; the solution I choose was perfectly adequate, and in his comment Kevin derided it without suggesting improvements or alternatives. This was not feedback, it was an exploitation of his position of control to obstruct my attempt to fix the issue. He also cited angry messages, which you can read in full here: https://pastebin.com/riVtbXAR

Thanks for reading. I hope that you will understand my desire to keep this project stable. I genuinely consider it to be some of the best work I did while at HumanGeo, and it is one of very few packages in any language that provide this particular functionality. I suppose the first thing I will do is add tests to guarantee correct behavior for unicode string inputs and a contributor code of conduct; hopefully I will be adding them to this repository and not to a restrictively-licensed fork.

**Comment by sfairgrieve** [comment]
@pineapplemachine I appreciate your continued passion for this project and desire to keep the project stable.  Yes, we accepted a PR without thoroughly testing it, so for that we apologize.  As Kevin mentioned, maintaining these repos is a side job among many side jobs that we voluntarily participate in, and we sometimes make mistakes.  Once you brought up the issue with the PR, we decided to pay closer attention to this repo and not just accept PRs with limited review, so as to prevent future mistakes.  

Personally, I don't believe Kevin abused his position of power.  He simply asked a relevant question that many Python project maintainers face at this point "is it worth maintaining support for Python 2 in addition to supporting Python 3?" and then provided some feedback on the fix that maintained Python 2 support.  In terms of the Python 2 question, it looks like it wasn't that hard to maintain Python 2 support, so that's great.  In terms of the feedback on the fix, I believe he was just looking for some comments explaining the `isinstance` check or a different solution that had a few more lines of code but was also potentially more clear to people reviewing the code.  There is a solution that doesn't require external libraries, it just requires a few more lines of code.  This type of feedback is not uncommon when people submit PRs; yes, it could have been worded differently or included more of an explanation, but I think there's room for everyone here to improve their word choices.

We're essentially arguing over some comments about a few characters of code at this point, and I don't think this needs to escalate into a hostile fork and re-license situation, but I can't keep you from doing that if that's your choice.  Unfortunately, I can't make the decision to give you write access at this time, but your question has prompted us to discuss this internally.

**Comment by pineapplemachine** [comment]
@sfairgrieve I appreciate that it is being discussed.

Unfortunately, Kevin did not specify what he was looking for. His comment derided my solution without offering improvements or alternatives. Moreover, by no means should it have been a blocking issue. The tests were failing for this repository for any version of Python, 2.x or 3.x until this PR was merged. A cleaner solution could have been discussed afterward. His intention was not to improve code quality, it was to obstruct my attempt to fix his mistake. If he was concerned with the quality or stability of this project, he would not have so readily merged the prior PR.

I will not compromise on the project which I consider to be my responsibility. If there is the opportunity for this to happen again with this original repository, then I will solve the problem by maintaining a fork instead. I recognize that Kevin's were the actions of one employee, but if I am not given write access then I will consider it to be an endorsement of his behavior; in this case my changes or additions will be licensed such that HumanGeo will not benefit from my future work on this package.
